quicklogic/qlf_k4n8/utils/create_lib.py

```diff
@@ -0,0 +1,416 @@
+                curr_str = "\n\t\tpin ({}) {}".format(
+                    pin.name, curly_braces_open
+                )
+                cap = "0.01686"
+                max_tran = "0.2000"
+                curr_str += form_pin_header(pin.dir, cap, max_tran)
+                curr_str += "\n\t\t{}".format(curly_braces_close)
+                lib_data += curr_str
+            else:
+                curr_str = "\n\t\tpin ({}) {}".format(
+                    pin.name, curly_braces_open
+                )
+                cap = "0.013646"
+                max_tran = "0.2000"
+                curr_str += form_pin_header(pin.dir, cap, max_tran)
+                curr_str += "\n\t\t{}".format(curly_braces_close)
+                lib_data += curr_str
+
+    dedicated_pin_lib_file = os.path.join(
+        curr_dir, '../devices/umc22/dedicated_pin_lib_data.txt'
+    )
+    dedicated_pin_data = open(dedicated_pin_lib_file, 'r').read()
+
+    inter_str = str1.replace("@dedicated_pin_data@", dedicated_pin_data)
+
+    final_str = inter_str.replace("@user_pin_data@", lib_data)
+    with open(lib_file_name, 'w') as out_fp:
+        out_fp.write(final_str)
+
+
+# =============================================================================
+
+
+def form_pin_header(direction, cap, max_tran):
+    '''
+    Form pin header section
+    '''
+    curr_str = "\n\t\t\tdirection : {};\n\t\t\tcapacitance : {};".format(
+        direction, cap
+    )
+    curr_str += "\n\t\t\tmax_transition : {};".format(max_tran)
+    return curr_str
+
+
+# =============================================================================
+
+
+def form_out_reset_timing_group(reset_name, timing_sense, timing_type):
+    '''
+    Form timing group for output pin when related pin is reset
+    '''
+    cell_fall_val = "4.6000"
+    fall_tran_val = "0.2000"
+    curly_braces_open = "{"
+    curly_braces_close = "}"
+    curr_str = "\n\t\t\ttiming () {}\n\t\t\t\trelated_pin : \"{}\";".format(
+        curly_braces_open, reset_name
+    )
+    curr_str += "\n\t\t\t\ttiming_sense : {};".format(timing_sense)
+    curr_str += "\n\t\t\t\ttiming_type : {};".format(timing_type)
+    curr_str += "\n\t\t\t\tcell_fall (scalar) {}\n\t\t\t\t\tvalues({});".format(
+        curly_braces_open, cell_fall_val
+    )
+    curr_str += "\n\t\t\t\t{}".format(curly_braces_close)
+    curr_str += "\n\t\t\t\tfall_transition (scalar) {}\n\t\t\t\t\tvalues({});".format(
+        curly_braces_open, fall_tran_val
+    )
+    curr_str += "\n\t\t\t\t{}".format(curly_braces_close)
+    curr_str += "\n\t\t\t{}".format(curly_braces_close)
+    return curr_str
+
+
+# =============================================================================
+
+
+def form_out_timing_group(clk_name, timing_type):
+    '''
+    Form timing group for output pin in a Pin group in a library file
+    '''
+    cell_rise_val = "2.59"
+    cell_fall_val = "2.59"
+    rise_tran_val = "0.85"
+    fall_tran_val = "0.85"
```

Comment:
There are lot of hard-coded values throughout the script. If they are hard constants common to all devices then this is fine. But if they are device-specific I'd add a mechanism to read them eg. from a JSON file (like the lib header template is). We can leave that as is if it is not critical for now.

Reply:
Incorporated this in the last checkin. Created a JSON file to pick these values.